### PR TITLE
Cleanup and preparatory work for adding 'open' declarations to information known at debug time

### PR DESCRIPTION
**Pull request by dsyme**
This is cleanup and preparatory work for adding 'open' declarations to information known at debug time
related to very long-standing bug #1003 

* [x] Collect information about "open" declarations during checking
* [x] Cleanup

Some issues arise when actually propagating this information into the PDB - see the comments at the end of the PR.  However the cleanup and information-plumbing is a good basis for future work on debugging (and I'd like to get it in as a basis for fixing other debugging bugs we have - it would be much easier not to handle the conflicts, and the code is now quite a lot cleaner)

NOTE: the original version of this PR also emitted the information into the PDB.

### CODE REVIEW

**Comment by dsyme** [comment]
The compiler part of this is working but populating the information in the PDB seems to cause problems in Visual Studio "watch" window.

I used the PPDBDump tool to dump the PDB contents for a one function program before and after this change. The d

Code:
```fsharp

open System

let f() = 
    Environment.HasShutdownStarted

f()
```

Before
```
C:\GitHub\dsyme\fsharp>C:\GitHub\dsyme\PPDB\bin\Dump\Debug\PPDBDump.exe a.pdb
empty

0x0000001d 0x00000001 0x00000024 0x00000002
Doc: C:\GitHub\dsyme\fsharp\a.fs
        Hash Algorithm: {8829d00f-11b8-4213-878b-770e8597ac16}
        Language: {ab4f38c9-b6e6-43ba-be3b-58080b2ccce3}

0x00000001 0x00000045
Debug: Doc Idx: 1
        Sequence Points:
                0 5 5 5 35 1
0x00000001 0x0000004c
Debug: Doc Idx: 1
        Sequence Points:
                0 7 1 7 4 1

empty

empty

0x00000001 0x00000000 0x00000001 0x00000000 0x00000000 0x00000006
LocalScope: 1 0 1 0 0 6
0x00000002 0x00000000 0x00000001 0x00000000 0x00000000 0x00000007
LocalScope: 2 0 1 0 0 7

empty

empty

empty
```

After
```
C:\GitHub\dsyme\fsharp>C:\GitHub\dsyme\PPDB\bin\Dump\Debug\PPDBDump.exe a.pdb
empty

0x0000001d 0x00000001 0x00000024 0x00000002
Doc: C:\GitHub\dsyme\fsharp\a.fs
        Hash Algorithm: {8829d00f-11b8-4213-878b-770e8597ac16}
        Language: {ab4f38c9-b6e6-43ba-be3b-58080b2ccce3}

0x00000001 0x00000045
Debug: Doc Idx: 1
        Sequence Points:
                0 5 5 5 35 1
0x00000001 0x000000e8
Debug: Doc Idx: 1
        Sequence Points:
                0 7 1 7 4 1

empty

empty

0x00000001 0x00000003 0x00000001 0x00000000 0x00000000 0x00000006
LocalScope: 1 3 1 0 0 6
0x00000002 0x00000000 0x00000001 0x00000000 0x00000000 0x00000007
LocalScope: 2 0 1 0 0 7

0x00000000 0x00000000
Import: Parent Idx: 0
        Imports:
0x00000001 0x000000b3
Import: Parent Idx: 1
        Imports:
                Microsoft
                Microsoft.FSharp
                0x1b400001
                Microsoft.FSharp.Core
                Microsoft.FSharp.Collections
                Microsoft.FSharp.Control
                0x02400001
                0x02400002
                0x01400002
                0x1b400002
                0x02400002
0x00000002 0x000000e4
Import: Parent Idx: 2
        Imports:
                System

empty

empty
```
Here the "Import" tables are correctly laid down as far as I can see.  The single method `f` refers to one of the local scopes with the `3` here:
```
0x00000001 0x00000003 0x00000001 0x00000000 0x00000000 0x00000006
LocalScope: 1 3 1 0 0 6
```

However as soon as we step into the `f()` the watch window shows this:

image

There is no other information being emitted differently, and it only applied when we have a method with ImportScope set, so it looks like adding this information triggers this problem specifically for F#.

It may be either 
1. VS expects other information populated as well, or
2. Some modification is needed

**Comment by dsyme** [comment]
OK there are two issues

- Something is going wrong with the emit of the "import type" debug declarations that correspond to "open" on F# modules. Removing these removes the "Internal F# Compiler Error" messages.

- However the import scope for `open System` still doesn't cause a sample such as `Environment.HasShutdownStarted` to resolve in the watch window for example. It could be that we have to emit either the compiler flags or metadata references which allow reproducible debugging, see #12002.  This information is present in C# portable PDBs but not F# ones. For example a smalle C# program gives this extra information in the PDB:

```
0x00000027 0x00000003 0x00000045
Custom: Parent: 7
        Kind: {b5feec05-8cd0-4a83-96da-466284bb4bd8}
        Data: 218 bytes
0x00000027 0x00000004 0x00000121
Custom: Parent: 7
        Kind: {7e4d4708-096e-4c5c-aeda-cb10ba6a740d}
        Data: 1978 bytes
```

> the presence of the new data in the PDB should be enough. Compiler flags are stored at the module level in
> CustomDebugInformation with B5FEEC05-8CD0-4A83-96DA-466284BB4BD8, and metadata references
> are stored with 7E4D4708-096E-4C5C-AEDA-CB10BA6A740D

**Comment by dsyme** [comment]
@KevinRansom Ok, so this is harder than I expected. It might be we have to do some or all of

1. emit compilation options
2. emit metadata references
3. possible write an F# debug engine for VS

I'm not certain though.  It surprised me that emitting the import scopes wasn't enough - my working assumption was that F# was making use of the C# expression evaluator, and that emitting the import scopes would be enough to get that evaluator to respect `open System` etc.   However it might be we're just not setting up a full C# compilation context. The problem is, if we have to start emitting the compilation options, we should really go all the way and set up a proper **F#** compilation context for expression evaluation - otherwise the options we embed in the portable PDBs may be plain wrong.   Emitting metadata references would be ok (they are language neutral) but emitting incorrect compilation options into PDBs would be wrong

**Comment by dsyme** [comment]
@KevinRansom My suggestion is that we get the plumbing work in this PR "in the bank" but don't actually do the final emit in the PDB file.  That is, change this into a preparatory PR that gets the "open" declaration propagation logic into the compiler and through the compilation process.  We can then pick this up again to investigate further.

File: src/fsharp/FSharp.Compiler.Service/StandardError.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Could not execute because the application was not found or a compatible .NET SDK is not installed.
```
Comment (vzarytovskii) [inline src/fsharp/FSharp.Compiler.Service/StandardError.txt:1]: Are these files intented to be commited? Shall we add them to the .gitignore?

File: src/fsharp/FSharp.Compiler.Service/StandardOutput.txt [hunk 2]
```diff
@@ -1 +1,6 @@
-5.0.104
```
Comment (vzarytovskii) [inline src/fsharp/FSharp.Compiler.Service/StandardOutput.txt:1]: Are these files intented to be commited? Shall we add them to the .gitignore?

**Review by vzarytovskii — APPROVED** [review]

File: src/fsharp/FSharp.Compiler.Service/StandardError.txt (hunk 1 above)
Comment (dsyme) [reply src/fsharp/FSharp.Compiler.Service/StandardError.txt]: Removed the file, thanks (yes should add to

**Pull request merged**